curp/src/client_new/mod.rs

```diff
@@ -0,0 +1,259 @@
+                }
+                async move {
+                    let mut protocol_client = ProtocolClient::connect(addr).await.map_err(|e| {
+                        tonic::Status::cancelled(format!("cannot connect to addr, error: {e}"))
+                    })?;
+                    let mut req = tonic::Request::new(FetchClusterRequest::default());
+                    req.set_timeout(propose_timeout);
+                    let fetch_cluster_res = protocol_client.fetch_cluster(req).await?.into_inner();
+                    Ok::<FetchClusterResponse, tonic::Status>(fetch_cluster_res)
+                }
+            })
+            .collect();
+        let mut err = tonic::Status::invalid_argument("addrs is empty");
+        // find the first one return `FetchClusterResponse`
+        while let Some(r) = futs.next().await {
+            match r {
+                Ok(r) => {
+                    self.cluster_version = Some(r.cluster_version);
+                    if let Some(id) = r.leader_id {
+                        self.leader_state = Some((id, r.term));
+                    }
+                    self.all_members = Some(r.into_members_addrs());
+                    return Ok(self);
+                }
+                Err(e) => err = e,
+            }
+        }
+        Err(err)
+    }
+
+    /// Build the client
+    ///
+    /// # Errors
+    ///
+    /// Return `tonic::transport::Error` for connection failure.
+    #[inline]
+    pub async fn build<C: Command>(
+        self,
+    ) -> Result<impl ClientApi<Error = tonic::Status, Cmd = C>, tonic::transport::Error> {
+        let mut builder = UnaryBuilder::<P>::new(
+            self.all_members.unwrap_or_else(|| {
+                unreachable!("must set the initial members or discover from some endpoints")
+            }),
+            UnaryConfig::new_full(
+                *self.config.propose_timeout(),
+                *self.config.wait_synced_timeout(),
+            ),
+        );
+        if let Some(version) = self.cluster_version {
+            builder.set_cluster_version(version);
+        }
+        if let Some((id, server)) = self.local_server {
+            builder.set_local_server(id, server);
+        }
+        if let Some((id, term)) = self.leader_state {
+            builder.set_leader_state(id, term);
+        }
+        let unary = builder.build::<C>().await?;
+
+        let retry_config = if *self.config.use_backoff() {
+            RetryConfig::new_exponential(
+                *self.config.initial_retry_timeout(),
+                *self.config.max_retry_timeout(),
+                *self.config.retry_count(),
+                false,
+            )
+        } else {
+            RetryConfig::new_fixed(
+                *self.config.initial_retry_timeout(),
+                *self.config.retry_count(),
+                false,
+            )
+        };
+        let client = Retry::new(unary, retry_config);
+        Ok(client)
+    }
+}
```

Comment:
We could have a dual insurance mechanism.

Reply:
The construction process of `Unary` is suitable to be placed in a separate function. Some of the constructor parameters need to be processed, and it might be better to put them in a separate builder.

Reply:
I still can't understand the benefits of this, can you provide a concrete example?

Reply:
used in https://github.com/xline-kv/Xline/pull/585/files#diff-849b69e9d6b4f2aa4030a8ad213cae06998eef92a40061e98e2bc7050d64b269